fix: exceeding max open trades

uhliksk opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description

<!--- Describe your changes in detail -->

- Allow buy orders to be placed for already open trades but prevent opening new trades.
- If `Grid Trade #1` buy order is executed and max. open trades limit is reached, all other `Grid Trade #1` buy orders are cancelled. This will prevent overshooting the limit while it will allow to wait in multiple symbol for the first buy order to be triggered.

## Related Issue

<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->

#398, #457

## Motivation and Context

<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

This will solve the problem with more opened trades than allowed while also will fix the issue when buy orders in already open trades weren't allowed if the limit was exceeded.

## How Has This Been Tested?

<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

I tested changing the limit on the fly to open more trades and then limiting them to lower value. Then waiting if excess buy orders are cancelled and buy orders in already opened trades are allowed.

## Screenshots (if appropriate):

uhliksk commented:
@chrisleekr Please, don't merge yet as there is still an issue with less traded symbols. As the `executeTrailingTrade` needs to be executed to cancel the excess buy orders, sometimes those orders are not cancelled on time and with just next price change they are executed on Binance right before the bot had a chance to remove them. I'm working on commit to fire `executeTrailingTrade` when needed on those buy orders which have to be cancelled before there are any websocket data received from Binance.

chrisleekr commented:
@uhliksk 
Is this ready to review?

uhliksk commented:
@chrisleekr Not yet. I was busy and I didn't finished the commit which will help to cancel the buy order even when there is no price change since the max. open trade limit is reached.

uhliksk commented:
@habibalkhabbaz I'm modifying `app/cronjob/trailingTrade/step/ensure-grid-trade-order-executed.js`. I simply added `const queue = require('../../trailingTradeHelper/queue');` and then later in code I used `queue.executeFor(logger, symbol);`, but it's throwing `TypeError: queue.executeFor is not a function`. Am I overlooking something?

chrisleekr commented:
@uhliksk Ah, just to confirm, do you want to include this fix in the release? Then we can wait for the release.

uhliksk commented:
@chrisleekr Yes, I was just waiting until #464 is merged to replace my simple queue to the new queueing mechanism. I thought it'll be simple function replace but I have an issue with `queue.executeFor` and I don't know what I'm doing wrong right now.

habibalkhabbaz commented:
> @habibalkhabbaz I'm modifying `app/cronjob/trailingTrade/step/ensure-grid-trade-order-executed.js`. I simply added `const queue = require('../../trailingTradeHelper/queue');` and then later in code I used `queue.executeFor(logger, symbol);`, but it's throwing `TypeError: queue.executeFor is not a function`. Am I overlooking something?

Hello @uhliksk 
I couldn't see an issue with your code. Maybe your forgot to rebuild or so after migrating?

uhliksk commented:
@habibalkhabbaz Sorry, I forgot to push last commit.

uhliksk commented:
@habibalkhabbaz This is the line throwing the error.

https://github.com/chrisleekr/binance-trading-bot/blob/1fe9e7284a6564c79c9a214ec894a8caafcc27c8/app/cronjob/trailingTrade/step/ensure-grid-trade-order-executed.js#L262

uhliksk commented:
@chrisleekr If it's not throwing the error for you, then you can merge, but for me it's logging this when buy order is filled:

`{"name":"binance-api","version":"0.0.88","hostname":"2fb1d3018816","pid":46,"gitHash":"unspecified","level":50,"err":{"message":"p.executeFor is not a function","name":"TypeError","stack":"TypeError: p.executeFor is not a function\n    at /srv/dist/server.js:1:32112\n    at Array.map (<anonymous>)\n    at execute (/srv/dist/server.js:1:32088)\n    at async /srv/dist/server.js:1:20849\n    at async errorHandlerWrapper (/srv/dist/server.js:1:100025)\n    at async execute (/srv/dist/server.js:1:19521)"},"msg":"p.executeFor is not a function","time":"2022-08-14T08:41:25.672Z","v":0}`

habibalkhabbaz commented on app/cronjob/trailingTrade/step/ensure-grid-trade-order-executed.js:
```diff
@@ -254,6 +255,13 @@ const execute = async (logger, rawData) => {
       );
+
+      // Queue other symbols to check if max. open trade is reached
+      symbols.map(async symbolToQueue => {
```
`map` can be replaced with `forEach` and I think no need for `async` here because we are not expecting a result.

habibalkhabbaz commented:
> @chrisleekr If it's not throwing the error for you, then you can merge, but for me it's logging this when buy order is filled:
> 
> `{"name":"binance-api","version":"0.0.88","hostname":"2fb1d3018816","pid":46,"gitHash":"unspecified","level":50,"err":{"message":"p.executeFor is not a function","name":"TypeError","stack":"TypeError: p.executeFor is not a function\n at /srv/dist/server.js:1:32112\n at Array.map (<anonymous>)\n at execute (/srv/dist/server.js:1:32088)\n at async /srv/dist/server.js:1:20849\n at async errorHandlerWrapper (/srv/dist/server.js:1:100025)\n at async execute (/srv/dist/server.js:1:19521)"},"msg":"p.executeFor is not a function","time":"2022-08-14T08:41:25.672Z","v":0}`

@uhliksk 
It seems that the reported error not exists with latest commits because I see the reported error for `p.executeFor` but your code is fine.
I just added one comment.

uhliksk commented:
@chrisleekr 
You're right, thank you. I replaced `map` with `forEach` and removed `async`. I'm waiting for buy order to be executed to see if the error will be logged again.

habibalkhabbaz commented:
That's great @uhliksk
If no more issues I think this PR will be ready to be reviewed by @chrisleekr 
Do you agree @uhliksk?

uhliksk commented:
@habibalkhabbaz Maybe. It's still throwing `p.executeFor is not a function` to me, but I'm also getting another error  in another part of code `Error: Illegal characters found in parameter 'symbol'`. Is it possible I have corrupted database? I'll check. If there is something corrupted then the error is probably problem in my environment only and the code itself is fine for review.

uhliksk commented:
@habibalkhabbaz Another observation, the error is logged single time when buy order is executed, but forEach will run it multiple times for all symbols so it looks like just a single symbol is throwing the error. That's should proof it's just my local problem and the code is really good to go.

uhliksk commented:
The `Illegal characters found` is from `node_modules/binance-api-node/dist/http-client.js`. It's not related. I'm sorry I'm throwing meaningless assumptions here. I'm just frustrated because I don't see why this `p.executeFor` error is occuring.

habibalkhabbaz commented:
@uhliksk 
To be honest, I am not sure from where the issue is coming.
Can you log `symbols` just before `forEach` to see what's happening?

uhliksk commented:
@habibalkhabbaz

Before the `forEach`:
`{"name":"binance-api","version":"0.0.88","hostname":"b4e46d6ba733","pid":46,"gitHash":"unspecified","server":"binance","helper":"queue","jobName":"trailingTrade","correlationId":"df54aa42-8ae0-4e41-8e77-6143e80af09f","symbol":"ATOMBUSD","stepName":"ensure-grid-trade-order-executed","level":50,"symbols":["BTCBUSD","ETHBUSD","XRPBUSD","ADABUSD","SOLBUSD","DOGEBUSD","DOTBUSD","TRXBUSD","MATICBUSD","SHIBBUSD","AVAXBUSD","LTCBUSD","UNIBUSD","ETCBUSD","FTTBUSD","LINKBUSD","NEARBUSD","XMRBUSD","ATOMBUSD","XLMBUSD","BCHBUSD","ALGOBUSD","APEBUSD","FLOWBUSD","VETBUSD","ICPBUSD","MANABUSD","SANDBUSD","XTZBUSD","HBARBUSD","THETABUSD","EGLDBUSD","AXSBUSD","QNTBUSD","AAVEBUSD","EOSBUSD","HNTBUSD","MKRBUSD","ZECBUSD","IOTABUSD","FTMBUSD","RUNEBUSD","FILBUSD","GRTBUSD","CHZBUSD","KLAYBUSD","XECBUSD","NEOBUSD","LDOBUSD","CRVBUSD","WAVESBUSD","BATBUSD","STXBUSD","ENJBUSD","CAKEBUSD","ZILBUSD","PAXGBUSD","LRCBUSD","DASHBUSD","GMTBUSD"],"msg":">>> Ready to start queueing","time":"2022-08-14T12:55:28.119Z","v":0}
`

Inside the `forEach`:

`{"name":"binance-api","version":"0.0.88","hostname":"b4e46d6ba733","pid":46,"gitHash":"unspecified","server":"binance","helper":"queue","jobName":"trailingTrade","correlationId":"df54aa42-8ae0-4e41-8e77-6143e80af09f","symbol":"ATOMBUSD","stepName":"ensure-grid-trade-order-executed","level":50,"symbols":["BTCBUSD","ETHBUSD","XRPBUSD","ADABUSD","SOLBUSD","DOGEBUSD","DOTBUSD","TRXBUSD","MATICBUSD","SHIBBUSD","AVAXBUSD","LTCBUSD","UNIBUSD","ETCBUSD","FTTBUSD","LINKBUSD","NEARBUSD","XMRBUSD","ATOMBUSD","XLMBUSD","BCHBUSD","ALGOBUSD","APEBUSD","FLOWBUSD","VETBUSD","ICPBUSD","MANABUSD","SANDBUSD","XTZBUSD","HBARBUSD","THETABUSD","EGLDBUSD","AXSBUSD","QNTBUSD","AAVEBUSD","EOSBUSD","HNTBUSD","MKRBUSD","ZECBUSD","IOTABUSD","FTMBUSD","RUNEBUSD","FILBUSD","GRTBUSD","CHZBUSD","KLAYBUSD","XECBUSD","NEOBUSD","LDOBUSD","CRVBUSD","WAVESBUSD","BATBUSD","STXBUSD","ENJBUSD","CAKEBUSD","ZILBUSD","PAXGBUSD","LRCBUSD","DASHBUSD","GMTBUSD"],"symbolToQueue":"BTCBUSD","msg":">>> Add to queue BTCBUSD","time":"2022-08-14T12:55:28.119Z","v":0}
`

Right after the `logger` there is the `executeFor`:

`{"name":"binance-api","version":"0.0.88","hostname":"b4e46d6ba733","pid":46,"gitHash":"unspecified","server":"binance","helper":"queue","jobName":"trailingTrade","correlationId":"df54aa42-8ae0-4e41-8e77-6143e80af09f","symbol":"ATOMBUSD","level":50,"err":{"message":"p.executeFor is not a function","name":"TypeError","stack":"TypeError: p.executeFor is not a function\n    at /srv/dist/server.js:1:32223\n    at Array.forEach (<anonymous>)\n    at execute (/srv/dist/server.js:1:32139)\n    at async /srv/dist/server.js:1:20849\n    at async errorHandlerWrapper (/srv/dist/server.js:1:100137)\n    at async execute (/srv/dist/server.js:1:19521)"},"debug":true,"saveLog":true,"msg":"⚠ Execution failed.","time":"2022-08-14T12:55:28.119Z","v":0}
`

It just don't see the function at all no matter what the parameters are.

uhliksk commented:
@habibalkhabbaz I even tried to log `queue` but I'm not sure if ít is supposed to return anything.

`{"name":"binance-api","version":"0.0.88","hostname":"7581d55b7dcb","pid":46,"gitHash":"unspecified","server":"binance","helper":"queue","jobName":"trailingTrade","correlationId":"f43cdd2a-2988-41bd-a367-06f229962dbd","symbol":"SHIBBUSD","stepName":"ensure-grid-trade-order-executed","level":50,"queue":{},"msg":">>> What is queue?","time":"2022-08-14T13:03:37.139Z","v":0}
`

uhliksk commented:
@habibalkhabbaz I tried to log `moment` to compare the result. If I compare the results there was empty array `queue` loged previously but in `moment` there is nothing logged at all. Does that mean the `queue` is initialized as empty array instead of object pointing to functions? Why `require` is not initializing it properly?

`{"name":"binance-api","version":"0.0.88","hostname":"c34ea28efdd5","pid":46,"gitHash":"unspecified","server":"binance","helper":"queue","jobName":"trailingTrade","correlationId":"a6b0569a-8c55-42f1-a62f-89c281bd5044","symbol":"LDOBUSD","stepName":"ensure-grid-trade-order-executed","level":50,"msg":">>> What is moment?","time":"2022-08-14T13:10:55.081Z","v":0}
`

habibalkhabbaz commented:
@uhliksk 
Okay, I think I found the issue.
The `queue` cannot be referenced inside the trailing trade steps.
It should be in anywhere else.
For example, add helper function in `common` file to add a symbol to queue from trailing trade.

What I prefer is, we can check the open orders and cancel them directly if we exceeded the max open orders instead of adding all symbols to the queue. It would be faster and we will have better performance.

What do you think?

uhliksk commented:
@habibalkhabbaz Do you mean checking the open orders in `ensure-grid-trade-order-executed.js` where I have `executeFor` right now?

habibalkhabbaz commented:
@uhliksk
You are right.

So, here is a draft of what I suggest.

After an order is executed it should do the following.

```
if (exceeded the max open trade) {
    // 1. Retrieve the open orders from cache
    // 2. Loop through them
    // 3. Cancel the order
}
```

By doing this it will call the queue automatically from the WebSocket updates.

uhliksk commented:
@habibalkhabbaz Ok, I'll rework the thing. Maybe I should also move the `cancelOrder` from `handle-open-orders` to `common` to not duplicate the same function here.

habibalkhabbaz commented:
@uhliksk 
Yes.
Let me know if you need any help in the implementation, I will be happy to work on this too.

uhliksk commented:
@habibalkhabbaz Yes, you can rework the tests after I moved `cancelOrder` to `common` if you like. Thank you.

uhliksk commented:
@habibalkhabbaz You can actually move the `cancelOrder` by yourself as I can't commit the change with the tests failing.

habibalkhabbaz commented:
> @habibalkhabbaz You can actually move the `cancelOrder` by yourself as I can't commit the change with the tests failing.

Okay, let me work on this.

habibalkhabbaz commented:
@uhliksk 
Done. I moved `cancelOrder` to common and I updated `handle-open-orders` accordingly.

habibalkhabbaz commented:
@uhliksk 
I think you have to merge the last commit by @chrisleekr from the master branch.

uhliksk commented:
@habibalkhabbaz Wow, you are pretty fast. Thank you.

uhliksk commented:
@habibalkhabbaz Ouch, you also removed the part of code blocking new buy orders when max. open trade limit is reached. :)

uhliksk commented on app/cronjob/trailingTrade/step/handle-open-orders.js:
```diff
@@ -158,17 +123,6 @@ const execute = async (logger, rawData) => {
             moment().toISOString()
           );
         }
-      } else if (await isExceedingMaxOpenTrades(logger, data)) {
```
@habibalkhabbaz This part is necessary to prevent new buy orders to be placed. It took me almost 30 minutes to recognize why I'm getting buy orders over the limit.

uhliksk commented:
@habibalkhabbaz @chrisleekr

Do you think we should allow manual buy orders above the max. open trades limit?

habibalkhabbaz commented:
> @habibalkhabbaz Ouch, you also removed the part of code blocking new buy orders when max. open trade limit is reached. :)

Opss. Sorry.
Do we really need this?

If we implement the above suggestion it should work as expected but I didn't implement it yet.
It should be inside `ensure-grid-trade-order-executed` as we agree.

> After an order is executed it should do the following.
>
> ```
> if (exceeded the max open trades) {
>     // 1. Retrieve the open orders from the cache
>     // 2. Loop through them
>     // 3. Cancel the order
> }
> ```
> By doing this it will call the queue automatically from the WebSocket updates.

Do you like to work on this? If not I can make the changes for sure.

uhliksk commented:
@habibalkhabbaz Checking after the order is executed is not enough as the order executed can be the exact order we want to cancel. We need to prevent this order to be placed even before there is buy order executed. That's why the check in `handle-open-orders` is also necessary.  But if we don't want to cancel manual orders above the max. open trades limit then we should not even place new orders automatically if max. open trade limit is exceeded.

uhliksk commented:
@habibalkhabbaz 
Just one modification:

`// 3. Cancel the order if placed in "Grid Trade #1".`

We don't want to cancel all orders because orders in already opened trades are not incrementing number of open trades thus we want to keep them.

I made the thing but as I was checking orders after execution and you removed the part in `handle-open-orders` if the buy order exceeding the limit was executed then it was just to late to cancel the order and that's why I was getting orders over limit.

That's why I'm asking if we want to cancel manual buy order because current implementation will not allow even the manual orders over max. open trades limit.

habibalkhabbaz commented:
@uhliksk 
Yes, I understand what you want do now.
Let me fix this.

uhliksk commented:
@habibalkhabbaz Ok, I'll commit the changes I've made. I returned the code in `handle-open-orders` back so if you have nicer solution then you can fix. Thank you.

habibalkhabbaz commented:
@uhliksk 
I didn't touch `handle-open-orders`. Your implementation will be good.
I am working on `ensure-grid-trade-order-executed` to push only the symbols that have open orders to the queue.

uhliksk commented:
@habibalkhabbaz 
Ok, then I'll not commit anything and I'll wait for your solution. Thank you.

uhliksk commented:
> @uhliksk I didn't touch `handle-open-orders`.

I mean this part:
![image](https://user-images.githubusercontent.com/37454226/184551429-e90e312e-86b4-4701-b1eb-768b309ab2a0.png)

habibalkhabbaz commented:
@uhliksk 
Yes, I'll put it back. Work in progress :)

uhliksk commented:
@habibalkhabbaz
Excelent! You combined it with original approach. It's much simpler than what I was trying to do by cancelling the orders right in the `ensure-grid-trade-order-executed`.

habibalkhabbaz commented:
@uhliksk 
You did already a great job in this PR.
I just made the changes in `ensure-grid-trade-order-executed` which will push only symbols that have open orders to the queue.
Try it :)

uhliksk commented:
@habibalkhabbaz
Bad news.

`{"name":"binance-api","version":"0.0.88","hostname":"9a8f2c00b037","pid":46,"gitHash":"unspecified","server":"binance","helper":"queue","jobName":"trailingTrade","correlationId":"1c9e262b-47b1-4eb0-9dce-1e5d392bde17","symbol":"ETHBUSD","level":50,"err":{"message":"u.executeFor is not a function","name":"TypeError","stack":"TypeError: u.executeFor is not a function\n    at /srv/dist/server.js:1:75727\n    at Array.forEach (<anonymous>)\n    at checkIfOpenOrdersExceedingMaxOpenTrades (/srv/dist/server.js:1:75713)\n    at runMicrotasks (<anonymous>)\n    at processTicksAndRejections (internal/process/task_queues.js:95:5)\n    at async execute (/srv/dist/server.js:1:31914)\n    at async /srv/dist/server.js:1:20849\n    at async errorHandlerWrapper (/srv/dist/server.js:1:99150)\n    at async execute (/srv/dist/server.js:1:19521)"},"debug":true,"saveLog":true,"msg":"⚠ Execution failed.","time":"2022-08-14T20:12:42.038Z","v":0}
`

uhliksk commented:
@habibalkhabbaz 

This is what I was trying in `ensure-grid-trade-order-executed` but also didn't worked as expected.

```
      symbols.forEach(async symbolToCheck => {
        if (symbolToCheck !== symbol) {
          const orderData = {
            symbol,
            symbolConfiguration: {},
            openOrders: [],
            sell: {}
          };
          orderData.symbol = symbolToCheck;
          orderData.symbolConfiguration = await getConfiguration(
            logger,
            orderData.symbol
          );
          const lastBuyPriceDoc = getLastBuyPrice(logger, orderData.symbol);
          orderData.sell = {
            lastBuyPrice: _.get(lastBuyPriceDoc, 'lastBuyPrice', null)
          };

          // Only the initial buy orders have to be cancelled
          if (!orderData.sell.lastBuyPrice) {
            orderData.openOrders =
              JSON.parse(
                await cache.hget('trailing-trade-open-orders', orderData.symbol)
              ) || [];

            orderData.openOrders.forEach(order => {
              if (
                order.type === 'STOP_LOSS_LIMIT' &&
                order.side.toLowerCase() === 'buy' &&
                isExceedingMaxOpenTrades(logger, orderData)
              ) {
                // Cancel the buy order if max. open trades exceeded
                orderData.action = 'buy-order-cancelled';
                logger.info(
                  { orderData, saveLog: true },
                  `The current number of open trades has reached the maximum number of open trades. ` +
                    `The buy order will be cancelled.`
                );

                // Cancel current order
                cancelOrder(logger, orderData.symbol, order);
              }
            });
          }
        }
      });
```

habibalkhabbaz commented:
> @habibalkhabbaz Bad news.
> 
> `{"name":"binance-api","version":"0.0.88","hostname":"9a8f2c00b037","pid":46,"gitHash":"unspecified","server":"binance","helper":"queue","jobName":"trailingTrade","correlationId":"1c9e262b-47b1-4eb0-9dce-1e5d392bde17","symbol":"ETHBUSD","level":50,"err":{"message":"u.executeFor is not a function","name":"TypeError","stack":"TypeError: u.executeFor is not a function\n at /srv/dist/server.js:1:75727\n at Array.forEach (<anonymous>)\n at checkIfOpenOrdersExceedingMaxOpenTrades (/srv/dist/server.js:1:75713)\n at runMicrotasks (<anonymous>)\n at processTicksAndRejections (internal/process/task_queues.js:95:5)\n at async execute (/srv/dist/server.js:1:31914)\n at async /srv/dist/server.js:1:20849\n at async errorHandlerWrapper (/srv/dist/server.js:1:99150)\n at async execute (/srv/dist/server.js:1:19521)"},"debug":true,"saveLog":true,"msg":"⚠ Execution failed.","time":"2022-08-14T20:12:42.038Z","v":0} `

Ops.
I am running my bot on your branch right now. Let me see if I will face the same error.

habibalkhabbaz commented:
> @habibalkhabbaz
> 
> This is what I was trying in `ensure-grid-trade-order-executed` but also didn't worked as expected.
> 
> ```
>       symbols.forEach(async symbolToCheck => {
>         if (symbolToCheck !== symbol) {
>           const orderData = {
>             symbol,
>             symbolConfiguration: {},
>             openOrders: [],
>             sell: {}
>           };
>           orderData.symbol = symbolToCheck;
>           orderData.symbolConfiguration = await getConfiguration(
>             logger,
>             orderData.symbol
>           );
>           const lastBuyPriceDoc = getLastBuyPrice(logger, orderData.symbol);
>           orderData.sell = {
>             lastBuyPrice: _.get(lastBuyPriceDoc, 'lastBuyPrice', null)
>           };
> 
>           // Only the initial buy orders have to be cancelled
>           if (!orderData.sell.lastBuyPrice) {
>             orderData.openOrders =
>               JSON.parse(
>                 await cache.hget('trailing-trade-open-orders', orderData.symbol)
>               ) || [];
> 
>             orderData.openOrders.forEach(order => {
>               if (
>                 order.type === 'STOP_LOSS_LIMIT' &&
>                 order.side.toLowerCase() === 'buy' &&
>                 isExceedingMaxOpenTrades(logger, orderData)
>               ) {
>                 // Cancel the buy order if max. open trades exceeded
>                 orderData.action = 'buy-order-cancelled';
>                 logger.info(
>                   { orderData, saveLog: true },
>                   `The current number of open trades has reached the maximum number of open trades. ` +
>                     `The buy order will be cancelled.`
>                 );
> 
>                 // Cancel current order
>                 cancelOrder(logger, orderData.symbol, order);
>               }
>             });
>           }
>         }
>       });
> ```

@uhliksk That's great but I think it will be complicated and it will be hard to debug.
I will try to find a solution for that issue.

uhliksk commented:
@habibalkhabbaz 
Will it help to split `queue` into `queue-init` and `queue-executeFor` to not have `const { executeTrailingTrade } = require('../index');` included in `executeFor` part and thus workaround the issue?

habibalkhabbaz commented:
@uhliksk 
Yes, the same problem as I expected. The issue comes from circular dependency.
We cannot reference the queue inside the steps because the queue has a reference to the trailing trade.

```
Warning: Accessing non-existent property 'executeFor' of module exports inside circular dependency
```

```
{
  "name": "binance-api",
  "version": "0.0.88",
  "hostname": "6ccb1958e63c",
  "pid": 65,
  "gitHash": "unspecified",
  "server": "binance",
  "helper": "queue",
  "jobName": "trailingTrade",
  "correlationId": "127742bd-01e5-45cc-a5d6-e6a1da135422",
  "symbol": "MANAUSDT",
  "level": 50,
  "err": {
    "message": "queue.executeFor is not a function",
    "name": "TypeError",
    "stack": "TypeError: queue.executeFor is not a function\n    at /srv/app/cronjob/trailingTradeHelper/common.js:1151:35\n    at Array.forEach (<anonymous>)\n    at checkIfOpenOrdersExceedingMaxOpenTrades (/srv/app/cronjob/trailingTradeHelper/common.js:1151:11)\n    at runMicrotasks (<anonymous>)\n    at processTicksAndRejections (internal/process/task_queues.js:95:5)\n    at async execute (/srv/app/cronjob/trailingTrade/step/ensure-grid-trade-order-executed.js:216:5)\n    at async /srv/app/cronjob/trailingTrade.js:156:14\n    at async errorHandlerWrapper (/srv/app/error-handler.js:71:5)\n    at async execute (/srv/app/cronjob/trailingTrade.js:37:3)"
  },
  "debug": true,
  "saveLog": true,
  "msg": "⚠ Execution failed.",
  "time": "2022-08-14T20:27:21.404Z",
  "v": 0
}
```

Maybe we have to refactor the queue as you said.

uhliksk commented:
> Maybe we have to refactor the queue as you said.

@habibalkhabbaz 
But I'm actually not sure how to do it properly. I'll keep that `queue` thing on you. ;)

uhliksk commented:
@habibalkhabbaz 
Another idea I have is to set `queued-symbol` into cache instead of calling `executeFor` directly and everytime `executeTrailingTrade` is finishing it'll look into cache if there are some `queued-symbol` keys and will call `executeFor` for those symbols and then will remove those keys from cache. What do you think?

habibalkhabbaz commented:
@uhliksk 
I use PubSub to do the trick. It should work properly now. Try and let me know :)
Maybe @chrisleekr has other thoughts but at least for now, this implementation does the trick with simplicity.

uhliksk commented:
@habibalkhabbaz 
It looks like it doesn't work. Orders has been cancelled on another tick in `handle-open-orders`. One of them even 6 minutes after UNIBUSD buy order has been executed.

![image](https://user-images.githubusercontent.com/37454226/184555598-de620352-6916-4281-b0c1-f8c60787067a.png)

habibalkhabbaz commented:
Hello @uhliksk 
Is this with the latest commit ?

uhliksk commented:
@habibalkhabbaz 
With last commit it looks working as expected.

![image](https://user-images.githubusercontent.com/37454226/184556211-e1bf6ddb-a0c7-490c-a408-ee3b98ad452a.png)

uhliksk commented:
Just a note: Orders are cancelled on Binance instantly, notifications in bot are triggered instantly, but from `Buy Open Orders` some of them are still not removed and are wating for next `Action`. It looks like order cancellation received by websocket is missing the call of `executeFor`.

uhliksk commented:
It's the same with manual orders. I'll create order, it will show on Binance and bot almost instantly. I'll cancel order, it is cancelled on Binance instantly, but in bot on another price change. Do you know @habibalkhabbaz where it should be fixed?

habibalkhabbaz commented:
> Just a note: Orders are cancelled on Binance instantly, notifications in bot are triggered instantly, but from `Buy Open Orders` some of them are still not removed and are wating for next `Action`. It looks like order cancellation received by websocket is missing the call of `executeFor`.

I pushed a fix for this.

> It's the same with manual orders. I'll create order, it will show on Binance and bot almost instantly. I'll cancel order, it is cancelled on Binance instantly, but in bot on another price change. Do you know @habibalkhabbaz where it should be fixed?

However, for manual order cancellation, it should be reflected on the frontend directly.

https://github.com/uhliksk/binance-trading-bot/blob/d964d22970acec275b382caf21f96f2190dea7af/app/frontend/websocket/handlers/cancel-order.js#L26

uhliksk commented:
> I pushed a fix for this.

I fixed a fix for this. :)

habibalkhabbaz commented:
@uhliksk @chrisleekr 
My bot running on this branch till now and it seems stable enough.

chrisleekr commented:
@habibalkhabbaz @uhliksk 

Sorry for joining the late party.

Let me review and merge in!

chrisleekr commented on app/cronjob/trailingTrade/step/ensure-grid-trade-order-executed.js:
```diff
@@ -248,6 +248,8 @@ const execute = async (logger, rawData) => {
         temporaryDisableActionAfterConfirmingOrder
       );
+
+      PubSub.publish('check-open-orders', {});
```
I think we should add to SELL filled order to check open orders as well.
I will add in.

habibalkhabbaz replied:
Hello @chrisleekr 
I am not sure if we really need it in SELL filled order as well.
We add this only in BUY filled order to check the rest open order and cancel them if we reached the limit.

chrisleekr replied:
You are absolutely right.
Scratch my comment.

chrisleekr commented on app/cronjob/trailingTrade/step/handle-open-orders.js:
```diff
@@ -88,7 +55,23 @@ const execute = async (logger, rawData) => {
+
+        // Reset buy open orders
+        if (cancelResult === true) {
+          data.buy.openOrders = [];
```
Cancelling the order often fails even if it was successful. 

And when it succeeds to cancel, it should process the same as stop > limit price cancel process.

https://github.com/chrisleekr/binance-trading-bot/pull/462/files#diff-48b71d5ac0a2a4fd94708519e33e808e09b1a98ab8df1b8996f499d8b71a244bR124

Agree? @uhliksk @habibalkhabbaz

habibalkhabbaz replied:
I agree.

chrisleekr replied:
Let me quickly refactor it.

uhliksk replied:
I agree. It'll be cleaner way even as this is actually cancelling the first buy order which have no impact on anything. It will save some headache in future if it will be reused for another purpose.

chrisleekr replied:
I was gonna just update this file, but @uhliksk is right. I will make a generic function in the common.js and update other steps to reduce duplicated codes.

Allow me some time.

chrisleekr replied:
@uhliksk @habibalkhabbaz 

I've added a generic function `refreshOpenOrdersAndAccountInfo ` to `common.js` - https://github.com/chrisleekr/binance-trading-bot/pull/462/files#diff-478c565f3dc02bbdb099f674677ebcd7d1f16fa4ddd06220dd3d6119b8467b84R1130

And I updated `handle-open-orders.js` to use  `refreshOpenOrdersAndAccountInfo` - https://github.com/chrisleekr/binance-trading-bot/pull/462/files#diff-48b71d5ac0a2a4fd94708519e33e808e09b1a98ab8df1b8996f499d8b71a244bR127

I checked and found some other steps have exactly the same code.
However, refactoring those steps seems out of scope. Hence, I've added FIXME comments instead of refactoring them - https://github.com/chrisleekr/binance-trading-bot/pull/462/files#diff-ad373fc5bc2ac395a075a4ef44bdaebd17d66fc02707c5c07c83e1483199bab6R205

Let's test it out for another day and if there is no issue, I will merge in.

This is a very good fix @uhliksk @habibalkhabbaz

chrisleekr commented:
Let’s merge it!
